File: epi_judge_python/primitive_divide.py
from test_framework import generic_test


def divide(x: int, y: int) -> int:
    result = 0

    for power in range(31, -1, -1):
        y_power = y << power
        if x >= y_power:
            result += 1 << power
            x -= y_power

    return result


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('primitive_divide.py',
                                       'primitive_divide.tsv', divide))

File: epi_judge_cpp/primitive_divide.cc
#include "test_framework/generic_test.h"

int Divide(int x, int y) {
    int result = 0;

    int shift = 32;
    unsigned long long yPow = static_cast<unsigned long long>(y) << shift;

    while (x >= y) {
        while (yPow > x) {
            yPow >>= 1;
            --shift;
        }

        result += 1LL << shift;
        x -= yPow;
    }

    return result;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"x", "y"};
    return GenericTestMain(args, "primitive_divide.cc", "primitive_divide.tsv",
                           &Divide, DefaultComparator{}, param_names);
}
